### ISSUE TITLE
Segmentation fault while indexing a list

### ISSUE THREAD

**Issue by jangedoo** [duckdb/duckdb#5020]
### What happens?

When indexing a list, segmentation fault occurs.  I have a database with a table called `test`.

```python
import duckdb
con2 = duckdb.connect("segfault.duckdb")
print(con2execute("describe test").df())
```
```
column_name column_type null  key default  extra
         ids   VARCHAR[]  YES  NaN     NaN    NaN
```

Btw, this not only occurs on Python but also when using DBeaver to connect and query the database.

### To Reproduce

```python
import duckdb
con2 = duckdb.connect("segfault.duckdb")
con2.execute("select ids[1] as id from test").df() # this causes the error
```

Sample database:
[segfault.duckdb.zip](https://github.com/duckdb/duckdb/files/9812004/segfault.duckdb.zip)

### OS:

Mac os monterey 12.5.1

### DuckDB Version:

0.5.1

### DuckDB Client:

Python 3.8

### Full Name:

Sanjaya Subedi

### Affiliation:

none

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by Tishj** [comment]
I'm not really sure if this is reproducible enough, because I can verify that I get the same result from running that query on the dataset you provided.
But a simple
```sql
select ids from test;
```
already triggers an assertion in `list_column_data.cpp:116`
```cpp
	if (child_scan_count > 0) {
		auto &child_entry = ListVector::GetEntry(result);
		D_ASSERT(child_entry.GetType().InternalType() == PhysicalType::STRUCT ||
		         state.child_states[1].row_index + child_scan_count <=
		             child_column->start + child_column->GetMaxEntry());
		child_column->ScanCount(state.child_states[1], child_entry, child_scan_count);
	}
```

**Comment by Tishj** [comment]
So I think the issue here is somewhere a malformed List gets created, and the issue isn't necessarily in list_extract

**Comment by Mytherin** [comment]
Thanks for the report! This should also be fixed by #5151. 



> I'm not really sure if this is reproducible enough, because I can verify that I get the same result from running that query on the dataset you provided. But a simple
> 
> ```sql
> select ids from test;
> ```
> 
> already triggers an assertion in `list_column_data.cpp:116`
> 
> ```c++
> 	if (child_scan_count > 0) {
> 		auto &child_entry = ListVector::GetEntry(result);
> 		D_ASSERT(child_entry.GetType().InternalType() == PhysicalType::STRUCT ||
> 		         state.child_states[1].row_index + child_scan_count <=
> 		             child_column->start + child_column->GetMaxEntry());
> 		child_column->ScanCount(state.child_states[1], child_entry, child_scan_count);
> 	}
> ```
> 
>  ```
> (lldb) p state.child_states[1].row_index + child_scan_count 
> (unsigned long long) $6 = 251278
> ```
> 
>  ```
> (lldb) p child_column->start + child_column->GetMaxEntry()
> (unsigned long long) $8 = 250480
> ```
> 
> Doesn't this mean that the list entries span multiple row groups? These could be separate issues, but I'm inclined to think the list_extract issue is a result of this issue. If we believe we have scanned all the child values of the list, but we didn't because they weren't contained in the rowgroup we were scanning.
> 
> Then that results in a read of uninitialized memory/heap-buffer overflow when we take the `list_entry.length` as our truth value.

This issue is caused by an incompatibility in the storage format between the current master and 0.5.1. However, the storage version was not correctly incremented.  Essentially the way that list offsets are stored was changed to allow for merging row groups and to avoid storing high numbers unnecessarily (https://github.com/duckdb/duckdb/commit/a17c41cf2026620e0eab787cf189009bb0ce64bf).
